File: placesearch/src/main/java/org/tat/sdksample/manager/PlaceDetailManager.java
package org.tat.sdksample.manager;

import org.tat.sdksample.inteface.DetailLoadingListener;
import org.tat.sdksample.model.Detail;
import org.th.tatsdk.ServiceRequestListener;
import org.th.tatsdk.common.TATItemCode;
import org.th.tatsdk.common.TATLanguage;
import org.th.tatsdk.common.TATLocation;
import org.th.tatsdk.common.TATPlaceContact;
import org.th.tatsdk.search.TATGetPlaceDetailParameter;
import org.th.tatsdk.search.accommodation.TATGetAccommodationDetail;
import org.th.tatsdk.search.accommodation.TATGetAccommodationDetailResult;
import org.th.tatsdk.search.attraction.TATGetAttractionDetail;
import org.th.tatsdk.search.attraction.TATGetAttractionDetailResult;
import org.th.tatsdk.search.other.TATGetOtherPlaceDetail;
import org.th.tatsdk.search.other.TATGetOtherPlaceDetailResult;
import org.th.tatsdk.search.restaurant.TATGetRestaurantDetail;
import org.th.tatsdk.search.restaurant.TATGetRestaurantDetailResult;
import org.th.tatsdk.search.shop.TATGetShopDetail;
import org.th.tatsdk.search.shop.TATGetShopDetailResult;

public class PlaceDetailManager {

    DetailLoadingListener listener = null;

    public PlaceDetailManager(DetailLoadingListener listener) {
        this.listener = listener;
    }

    /**
     * @loadPlaceDetail request for detail of place using place id for each category.
     * For this example will map only used value.
     */
    public void loadPlaceDetail(String placeId, String placeCategory) {

        TATGetPlaceDetailParameter param = new TATGetPlaceDetailParameter(placeId, TATLanguage.ENGLISH);

        switch (placeCategory) {
            case "Attraction":
                TATGetAttractionDetail.executeAsync(param, new ServiceRequestListener<TATGetAttractionDetailResult>() {
                    @Override
                    public void onResponse(TATGetAttractionDetailResult result) {
                        TATLocation location = result.getLocation();
                        Detail detail = new Detail(result.getName(),
                                Detail.sortAddress(location.getAddress(), location.getSubDistrict(), location.getDistrict(), location.getProvince()),
                                result.getInfo().getDetail(),
                                checkTelephone(result.getContact()),
                                checkUrl(result.getContact()),
                                sortList(result.getFacilities()),
                                sortList(result.getServices()),
                                sortList(result.getPayments())
                        );
                        if (listener != null) {
                            listener.onSucceed(detail);
                        }
                    }

                    @Override
                    public void onError(String message, int errorCode) {
                        if (listener != null) {
                            listener.onError(message, errorCode);
                        }
                    }
                });
                break;
            case "Accommodation":
                TATGetAccommodationDetail.executeAsync(param, new ServiceRequestListener<TATGetAccommodationDetailResult>() {
                    @Override
                    public void onResponse(TATGetAccommodationDetailResult result) {
                        TATLocation location = result.getLocation();
                        Detail detail = new Detail(result.getName(),
                                Detail.sortAddress(location.getAddress(), location.getSubDistrict(), location.getDistrict(), location.getProvince()),
                                result.getInfo().getDetail(),
                                checkTelephone(result.getContact()),
                                checkUrl(result.getContact()),
                                sortList(result.getFacilities()),
                                sortList(result.getServices()),
                                sortList(result.getPayments())
                        );
                        if (listener != null) {
                            listener.onSucceed(detail);
                        }
                    }

                    @Override
                    public void onError(String message, int errorCode) {
                        if (listener != null) {
                            listener.onError(message, errorCode);
                        }
                    }
                });
                break;
            case "Restaurant":
                TATGetRestaurantDetail.executeAsync(param, new ServiceRequestListener<TATGetRestaurantDetailResult>() {
                    @Override
                    public void onResponse(TATGetRestaurantDetailResult result) {
                        TATLocation location = result.getLocation();
                        Detail detail = new Detail(result.getName(),
                                Detail.sortAddress(location.getAddress(), location.getSubDistrict(), location.getDistrict(), location.getProvince()),
                                result.getInfo().getDetail(),
                                checkTelephone(result.getContact()),
                                checkUrl(result.getContact()),
                                sortList(result.getFacilities()),
                                sortList(result.getServices()),
                                sortList(result.getPayments())
                        );
                        if (listener != null) {
                            listener.onSucceed(detail);
                        }
                    }

                    @Override
                    public void onError(String message, int errorCode) {
                        if (listener != null) {
                            listener.onError(message, errorCode);
                        }
                    }
                });
                break;
            case "Shop":
                TATGetShopDetail.executeAsync(param, new ServiceRequestListener<TATGetShopDetailResult>() {
                    @Override
                    public void onResponse(TATGetShopDetailResult result) {
                        TATLocation location = result.getLocation();
                        Detail detail = new Detail(result.getName(),
                                Detail.sortAddress(location.getAddress(), location.getSubDistrict(), location.getDistrict(), location.getProvince()),
                                result.getInfo().getDetail(),
                                checkTelephone(result.getContact()),
                                checkUrl(result.getContact()),
                                sortList(result.getFacilities()),
                                sortList(result.getServices()),
                                sortList(result.getPayments())
                        );
                        if (listener != null) {
                            listener.onSucceed(detail);
                        }
                    }

                    @Override
                    public void onError(String message, int errorCode) {
                        if (listener != null) {
                            listener.onError(message, errorCode);
                        }
                    }
                });
                break;
            case "Other":
                TATGetOtherPlaceDetail.executeAsync(param, new ServiceRequestListener<TATGetOtherPlaceDetailResult>() {
                    @Override
                    public void onResponse(TATGetOtherPlaceDetailResult result) {
                        TATLocation location = result.getLocation();
                        Detail detail = new Detail(result.getName(),
                                Detail.sortAddress(location.getAddress(), location.getSubDistrict(), location.getDistrict(), location.getProvince()),
                                result.getInfo().getDetail(),
                                checkTelephone(result.getContact()),
                                checkUrl(result.getContact()),
                                "",
                                "",
                                ""
                        );
                        if (listener != null) {
                            listener.onSucceed(detail);
                        }
                    }

                    @Override
                    public void onError(String message, int errorCode) {
                        if (listener != null) {
                            listener.onError(message, errorCode);
                        }
                    }
                });
                break;
            default:
                break;
        }
    }

    private String checkTelephone(TATPlaceContact contact) {
        if (contact != null) {
            if (contact.getPhones() != null) {
                if (contact.getPhones().length > 0) {
                    return contact.getPhones()[0];
                } else return "";
            } else return "";
        } else return "";
    }

    private String checkUrl(TATPlaceContact contact) {
        if (contact != null) {
            if (contact.getUrls() != null) {
                if (contact.getUrls().length > 0) {
                    return contact.getUrls()[0];
                } else return "";
            } else return "";
        } else return "";
    }

    private String sortList(TATItemCode[] list) {
        if (list != null) {
            if (list.length > 0) {
                int index = 0;
                String nameList = "";
                for (TATItemCode item : list) {
                    if (index != 0) {
                        nameList = nameList + "," + item.getDesc();
                        index++;
                    } else {
                        nameList = item.getDesc();
                        index++;
                    }

                }
                return nameList;
            } else {
                return "";
            }
        } else {
            return "";
        }
    }
}

File: app/src/main/java/org/tat/sdksample/placesearch/manager/PlaceDetailManager.kt
package org.tat.sdksample.placesearch.manager

import org.tat.sdksample.placesearch.inteface.DetailLoadingListener
import org.tat.sdksample.placesearch.model.Detail
import org.th.tatsdk.ServiceRequestListener
import org.th.tatsdk.common.TATLanguage
import org.th.tatsdk.common.TATPlaceContact
import org.th.tatsdk.search.*

class PlaceDetailManager(listener: DetailLoadingListener) {

    internal var listener: DetailLoadingListener? = null

    init {
        this.listener = listener
    }

    ///request for detail of place using place id for each category.
    fun loadPlaceDetail(placeId: String, placeCategory: String) {

        val param = TATGetPlaceDetailParameter(placeId, TATLanguage.ENGLISH)

        when (placeCategory) {
            "Attraction" -> TATPlaces.getAttractionAsync(param, object : ServiceRequestListener<TATGetAttractionDetailResult> {
                override fun onResponse(result: TATGetAttractionDetailResult?) {
                    result?.let {
                        val location = result.location
                        val detail = Detail(result.name,
                                Detail.sortAddress(location?.address, location?.subDistrict, location?.district, location?.province),
                                result.information?.detail ?: "",
                                checkTelephone(result.contact),
                                checkUrl(result.contact),
                                sortList(result.facilities),
                                sortList(result.services),
                                sortList(result.paymentMethods)
                        )
                        if (listener != null) {
                            listener!!.onSucceed(detail)
                        }
                    }
                }

                override fun onError(message: String?, errorCode: Int) {
                    if (listener != null) {
                        listener!!.onError(message ?: "", errorCode)
                    }
                }
            })
            "Accommodation" -> TATPlaces.getAccommodationAsync(param, object : ServiceRequestListener<TATGetAccommodationDetailResult> {
                override fun onResponse(result: TATGetAccommodationDetailResult?) {
                    result?.let {
                        val location = result.location
                        val detail = Detail(result.name,
                                Detail.sortAddress(location?.address, location?.subDistrict, location?.district, location?.province),
                                result.information?.detail ?: "",
                                checkTelephone(result.contact),
                                checkUrl(result.contact),
                                sortList(result.facilities),
                                sortList(result.services),
                                sortList(result.paymentMethods)
                        )
                        if (listener != null) {
                            listener!!.onSucceed(detail)
                        }
                    }
                }

                override fun onError(message: String?, errorCode: Int) {
                    if (listener != null) {
                        listener!!.onError(message ?: "", errorCode)
                    }
                }
            })
            "Restaurant" -> TATPlaces.getRestaurantAsync(param, object : ServiceRequestListener<TATGetRestaurantDetailResult> {
                override fun onResponse(result: TATGetRestaurantDetailResult?) {
                    result?.let {
                        val location = result.location
                        val detail = Detail(result.name,
                                Detail.sortAddress(location?.address, location?.subDistrict, location?.district, location?.province),
                                result.information?.detail ?: "",
                                checkTelephone(result.contact),
                                checkUrl(result.contact),
                                sortList(result.facilities),
                                sortList(result.services),
                                sortList(result.paymentMethods)
                        )
                        if (listener != null) {
                            listener!!.onSucceed(detail)
                        }
                    }

                }

                override fun onError(message: String?, errorCode: Int) {
                    if (listener != null) {
                        listener!!.onError(message ?: "", errorCode)
                    }
                }
            })
            "Shop" -> TATPlaces.getShopAsync(param, object : ServiceRequestListener<TATGetShopDetailResult> {
                override fun onResponse(result: TATGetShopDetailResult?) {
                    result?.let {
                        val location = result.location
                        val detail = Detail(result.name,
                                Detail.sortAddress(location?.address, location?.subDistrict, location?.district, location?.province),
                                result.information?.detail ?: "",
                                checkTelephone(result.contact),
                                checkUrl(result.contact),
                                sortList(result.facilities),
                                sortList(result.services),
                                sortList(result.paymentMethods)
                        )
                        if (listener != null) {
                            listener!!.onSucceed(detail)
                        }
                    }
                }

                override fun onError(message: String?, errorCode: Int) {
                    if (listener != null) {
                        listener!!.onError(message ?: "", errorCode)
                    }
                }
            })
            "Other" -> TATPlaces.getOtherPlaceAsync(param, object : ServiceRequestListener<TATGetOtherPlaceDetailResult> {
                override fun onResponse(result: TATGetOtherPlaceDetailResult?) {
                    result?.let {
                        val location = result.location
                        val detail = Detail(result.name,
                                Detail.sortAddress(location?.address, location?.subDistrict, location?.district, location?.province),
                                result.information?.detail ?: "",
                                checkTelephone(result.contact),
                                checkUrl(result.contact),
                                "",
                                "",
                                ""
                        )
                        if (listener != null) {
                            listener!!.onSucceed(detail)
                        }
                    }
                }

                override fun onError(message: String?, errorCode: Int) {
                    if (listener != null) {
                        listener!!.onError(message ?: "", errorCode)
                    }
                }
            })
            else -> {
            }
        }
    }

    private fun checkTelephone(contact: TATPlaceContact?): String {
        return contact?.phones?.getOrNull(0) ?: ""
    }

    private fun checkUrl(contact: TATPlaceContact?): String {
        return contact?.urls?.getOrNull(0) ?: ""
    }

    private fun sortList(list: Array<TATItem>?): String {
        val member = list?.map {
            it.description
        }

        return member?.let {
            it.joinToString(",")
        } ?: ""
//        if (list != null) {
//            if (list.isNotEmpty()) {
//                var index = 0
//                var nameList = ""
//                for (item in list) {
//                    if (index != 0) {
//                        nameList = nameList + "," + item.description
//                        index++
//                    } else {
//                        nameList = item.description
//                        index++
//                    }
//
//                }
//                return nameList
//            } else {
//                return ""
//            }
//        } else {
//            return ""
//        }
    }
}
